lib/Conversion/TorchToLinalg/Pooling.cpp

```diff
@@ -228,6 +228,149 @@ class ConvertAtenAdaptiveAvgPool2dOp
+                  ConversionPatternRewriter &rewriter) const override {
+    if (failed(verifyLinalgCompatibleTypes(op, rewriter)))
+      return failure();
+    Location loc = op->getLoc();
+    Value self = adaptor.self();
+
+    Type elementType = self.getType().cast<RankedTensorType>().getElementType();
+    if (!elementType.isa<mlir::FloatType>())
+      return op.emitError("unimplemented: non-floating point type");
+
+    // Pattern match against the op's original operands, because otherwise we
+    // will get the lowered version of the operands which is harder to pattern
+    // match.
+    SmallVector<int64_t, 2> kernelSizeInts;
+    if (!matchPattern(op.kernel_size(), m_TorchConstantIntList(kernelSizeInts)))
+      return rewriter.notifyMatchFailure(
+          op, "only support constant int kernel size");
+    SmallVector<int64_t, 2> strideInts;
+    if (!matchPattern(op.stride(), m_TorchConstantIntList(strideInts)))
+      return rewriter.notifyMatchFailure(op,
+                                         "only support constant int strides");
+    SmallVector<int64_t, 2> paddingInts;
+    if (!matchPattern(op.padding(), m_TorchConstantIntList(paddingInts)))
+      return rewriter.notifyMatchFailure(op,
+                                         "only support constant int paddings");
+
+    // The ceil_mode should be a constant `False`.
+    bool ceilMode;
+    if (!matchPattern(op.ceil_mode(), m_TorchConstantBool(&ceilMode))) {
+      return rewriter.notifyMatchFailure(op, "ceil_mode must be a constant");
+    } else if (ceilMode) {
+      return rewriter.notifyMatchFailure(
+          op, "unimplemented: ceil_mode is expected to be false");
+    }
+
+    // The count_include_pad should be a constant `True`.
+    bool countIncludePad;
+    if (!matchPattern(op.count_include_pad(),
+                      m_TorchConstantBool(&countIncludePad))) {
+      return rewriter.notifyMatchFailure(
+          op, "count_include_pad must be a constant");
+    } else if (!countIncludePad) {
+      return rewriter.notifyMatchFailure(
+          op, "unimplemented: count_include_pad is expected to be true");
+    }
+
+    // The divisor_override should be either None or a constant int.
+    int64_t divisorOverride;
+    if (!op.divisor_override().getType().isa<Torch::NoneType>() &&
+        (!matchPattern(op.divisor_override(),
+                       m_TorchConstantInt(&divisorOverride)))) {
+      return rewriter.notifyMatchFailure(
+          op, "divisor_override must be either None or a constant int");
+    }
+
+    SmallVector<int64_t, 4> paddingIncludingNC = {0, 0};
+    paddingIncludingNC.insert(paddingIncludingNC.end(), paddingInts.begin(),
+                              paddingInts.end());
+    Value paddedInput = torch_to_linalg::getPaddedTensor(op, rewriter, self,
+                                                         paddingIncludingNC);
+
+    Value N = getDimOp(rewriter, loc, self, 0);
```

Comment:
this line until line 337 are almost the same as `ConvertAtenMaxPool2dOp`. Can you refactor the code with a helper function?

Reply:
I have multiple PRs opened for the different pooling ops. I will refactor the code in the last remaining PR.

Reply:
It took a great effort to refactor the code for pooling ops. I hope it looks fine now.